File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/adapter/tasks/RemindersAdapter.java
package com.habitrpg.android.habitica.ui.adapter.tasks;

import android.graphics.Color;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;

import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.helpers.RemindersManager;
import com.habitrpg.android.habitica.models.tasks.RemindersItem;
import com.habitrpg.android.habitica.ui.helpers.ItemTouchHelperAdapter;
import com.habitrpg.android.habitica.ui.helpers.ItemTouchHelperViewHolder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by keithholliday on 5/31/16.
 */
public class RemindersAdapter extends RecyclerView.Adapter<RemindersAdapter.ItemViewHolder>
        implements ItemTouchHelperAdapter, RemindersManager.ReminderTimeSelectedCallback {

    private final List<RemindersItem> reminders = new ArrayList<>();
    private final String taskType;
    private RemindersManager remindersManager;

    public RemindersAdapter(List<RemindersItem> remindersInc, String taskType) {
        this.reminders.addAll(remindersInc);
        this.remindersManager = new RemindersManager(taskType);
        this.taskType = taskType;
    }

    @Override
    public ItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.reminder_item, parent, false);
        return new ItemViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ItemViewHolder holder, int position) {
        Date time = reminders.get(position).getTime();
        holder.reminderItemTextView.setText(remindersManager.reminderTimeToString(time));
        holder.hour = time.getHours();
        holder.minute = time.getMinutes();
    }

    public void addItem(RemindersItem item) {
        reminders.add(item);
        notifyItemInserted(reminders.size() - 1);
    }

    public List<RemindersItem> getRemindersItems() {
        return reminders;
    }

    @Override
    public int getItemCount() {
        return reminders.size();
    }


    @Override
    public void onItemDismiss(int position) {
        if (position >= 0 && position < reminders.size()) {
            reminders.get(position).async().delete();
            reminders.remove(position);
            notifyItemRemoved(position);
        }
    }

    @Override
    public void onItemMove(int fromPosition, int toPosition) {
        Collections.swap(reminders, fromPosition, toPosition);
        notifyItemMoved(fromPosition, toPosition);
    }

    @Override
    public void onReminderTimeSelected(@Nullable RemindersItem remindersItem) {
        if (remindersItem == null) {
            return;
        }
        for (int pos = 0; pos < reminders.size(); pos++) {
            if (remindersItem.getId().equals(reminders.get(pos).getId())) {
                reminders.set(pos, remindersItem);
                notifyItemChanged(pos);
                break;
            }
        }
    }

    class ItemViewHolder extends RecyclerView.ViewHolder implements
            ItemTouchHelperViewHolder, Button.OnClickListener {

        @BindView(R.id.item_edittext)
        EditText reminderItemTextView;

        @BindView(R.id.delete_item_button)
        Button deleteButton;

        int hour, minute;

        ItemViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            deleteButton.setOnClickListener(this);

            reminderItemTextView.setOnClickListener(v -> {
                RemindersItem reminder = reminders.get(getAdapterPosition());

                remindersManager.createReminderTimeDialog(RemindersAdapter.this, taskType, v.getContext(), reminder);
            });
        }

        @Override
        public void onItemSelected() {
            itemView.setBackgroundColor(Color.LTGRAY);
        }

        @Override
        public void onItemClear() {
            itemView.setBackgroundColor(0);
        }

        @Override
        public void onClick(View v) {
            if (v == deleteButton) {
                RemindersAdapter.this.onItemDismiss(getAdapterPosition());
            }
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/adapter/tasks/RemindersAdapter.kt
package com.habitrpg.android.habitica.ui.adapter.tasks

import android.graphics.Color
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.helpers.RemindersManager
import com.habitrpg.android.habitica.models.tasks.RemindersItem
import com.habitrpg.android.habitica.ui.helpers.ItemTouchHelperAdapter
import com.habitrpg.android.habitica.ui.helpers.ItemTouchHelperViewHolder
import com.habitrpg.android.habitica.ui.helpers.bindView
import java.util.*

/**
 * Created by keithholliday on 5/31/16.
 */
class RemindersAdapter(private val taskType: String) : androidx.recyclerview.widget.RecyclerView.Adapter<RemindersAdapter.ItemViewHolder>(), ItemTouchHelperAdapter, RemindersManager.ReminderTimeSelectedCallback {

    private val reminders = ArrayList<RemindersItem>()
    private val remindersManager: RemindersManager = RemindersManager(taskType)

    val remindersItems: List<RemindersItem>
        get() = reminders

    fun setReminders(reminders: List<RemindersItem>) {
        this.reminders.addAll(reminders)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.reminder_item, parent, false)
        return ItemViewHolder(view)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        val time = reminders[position].time
        holder.reminderItemTextView.setText(remindersManager.reminderTimeToString(time))
        @Suppress("DEPRECATION")
        holder.hour = time?.hours ?: 0
        @Suppress("DEPRECATION")
        holder.minute = time?.minutes ?: 0
    }

    fun addItem(item: RemindersItem) {
        reminders.add(item)
        notifyItemInserted(reminders.size - 1)
    }

    override fun getItemCount(): Int {
        return reminders.size
    }


    override fun onItemDismiss(position: Int) {
        if (position >= 0 && position < reminders.size) {
            reminders.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int) {
        Collections.swap(reminders, fromPosition, toPosition)
        notifyItemMoved(fromPosition, toPosition)
    }

    override fun onReminderTimeSelected(remindersItem: RemindersItem?) {
        if (remindersItem == null) {
            return
        }
        for (pos in reminders.indices) {
            if (remindersItem.id == reminders[pos].id) {
                reminders[pos] = remindersItem
                notifyItemChanged(pos)
                break
            }
        }
    }

    inner class ItemViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView), ItemTouchHelperViewHolder, View.OnClickListener {

        internal val reminderItemTextView: EditText by bindView(itemView, R.id.item_edittext)
        private val deleteButton: Button by bindView(itemView, R.id.delete_item_button)

        var hour: Int = 0
        var minute: Int = 0

        init {
            deleteButton.setOnClickListener(this)

            reminderItemTextView.setOnClickListener { v ->
                val reminder = reminders[adapterPosition]

                remindersManager.createReminderTimeDialog(this@RemindersAdapter, taskType, v.context, reminder)
            }
        }

        override fun onItemSelected() {
            itemView.setBackgroundColor(Color.LTGRAY)
        }

        override fun onItemClear() {
            itemView.setBackgroundColor(0)
        }

        override fun onClick(v: View) {
            if (v === deleteButton) {
                this@RemindersAdapter.onItemDismiss(adapterPosition)
            }
        }
    }
}
